Add DataSourceRef field to PVC spec

bswartz opened the pull request:
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

Modify the behavior of the AnyVolumeDataSource alpha feature gate to enable
a new field, DataSourceRef, rather than modifying the behavior of the
existing DataSource field. This allows addition Volume Populators in a way
that doesn't risk breaking backwards compatibility, although it will
result in eventually deprecating the DataSource field.

#### Which issue(s) this PR fixes:
Fixes #103275

#### Special notes for your reviewer:

A lot of test cases have been added, but I think more could be added. I'm going to keep working on those which the change itself is reviewed.

#### Does this PR introduce a user-facing change?
```release-note
Add DataSourceRef alpha field to PVC spec, which allows contents other than PVCs and VolumeSnapshots to be data sources.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators
```

bswartz committed:
Add DataSourceRef field to PVC spec

Modify the behavior of the AnyVolumeDataSource alpha feature gate to enable
a new field, DataSourceRef, rather than modifying the behavior of the
existing DataSource field. This allows addition Volume Populators in a way
that doesn't risk breaking backwards compatibility, although it will
result in eventually deprecating the DataSource field.

bswartz commented:
/assign @thockin

fedebongio commented:
/remove-sig api-machinery
/sig storage

bswartz commented:
/assign @saad-ali

thockin commented on pkg/apis/core/types.go:
```diff
@@ -450,13 +450,21 @@ type PersistentVolumeClaimSpec struct {
+	// core object) or a PersistentVolumeClaim object.
+	// When this field is specified, volume binding will only succeed if the type of
+	// the specified object matches some installed volume populator or dynamic
+	// provisioner.
```
I think you need to explain the special relationship to the DataSource field

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -492,13 +492,21 @@ type PersistentVolumeClaimSpec struct {
+	// volume is desired. This may be any local object from a non-empty API group (non
+	// core object) or a PersistentVolumeClaim object.
+	// When this field is specified, volume binding will only succeed if the type of
+	// the specified object matches some installed volume populator or dynamic
```
Same as above - explain the multi-field realtionship, pl3ase

thockin commented on pkg/registry/core/persistentvolumeclaim/storage/storage.go:
```diff
@@ -73,6 +77,43 @@ func (r *REST) ShortNames() []string {
+// applies on Get, Create, and Update, but we need to distinguish between them.
+//
+// This will be called on both PersistentVolumeClaim and PersistentVolumeClaimList types.
+func (r *REST) defaultOnRead(obj runtime.Object) {
```
![if-you-like-it-then-you-should-have-put-a-test-on-it](https://user-images.githubusercontent.com/5595220/124326500-545f0980-db3b-11eb-8794-00eb54cdc6c4.jpg)

bswartz replied:
Good point, I tested the underlying function but this layer deserves a few unit tests. E2E tests will help as well.

thockin commented on pkg/api/persistentvolumeclaim/util.go:
```diff
@@ -67,3 +68,20 @@ func dataSourceIsEnabled(pvcSpec *core.PersistentVolumeClaimSpec) bool {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.AnyVolumeDataSource) {
+		return
+	}
+	if pvcSpec.DataSource != nil && pvcSpec.DataSourceRef == nil {
```
This isn't clearly what the KEP calls for.  It may actually WORK but a reader of this will not comprehend it.

Where is the code that clears DataSource when it's not a PVC or Snapshot?  Should that be here?  Otherwise what happens if I send `{ DataSource = Invalid, DataSourceRef = nil }`?  Invalid will get copied to DSRef, then (somewhere) DS will be cleared, then DS and DSRef are not the same.  This converts case 5 (in KEP) into case 6.

Basically, the compat logic is so precise and subtle that I think you should centralize it and comment the heck out of it.  Take pity on the next poor sucker who deals with this.

bswartz replied:
So the case you mention is covered by the fact that DropDisabledFields() is called before NormalizeDataSources() in PrepareForCreate() in stragtegy.go. I agree it's not easy to follow, but there are substantial tests already for the existing code, so there's a strong reason not to make a ton of changes to the existing code, and to make these new changes in a mostly additive way. I can try to address the problem with lots and lots of comments.

thockin replied:
By convention, the DropDisabledFields pattern holds feature-gated logic.  No feature-gates == no logic.  I'd STRONGLY encourage you to gather all of the "permanent" logic together.  Having to jump between functions to convince myself this works is unpleasant at best (and the tests are all split) and very brittle.  I can EASILY see logic around here getting "fixed" wrongly because it's not at all clear how it is supposed to work or why.

thockin reviewed (commented):
I'll be online off and on today and tomorrow.  If you want to make an argument for "the tests work, let's merge it and then fix this" I am open to it, but I really REALLY think it should per made more permanent, since it is now part of the defined semantics.

bswartz commented:
@thockin Thanks I've been working on refactoring this code and adding more unit tests today. I'll send up another version trying to incorporate all your suggestions. The one tricky one will be the DropDisabledFields() implementation -- I understand it's intended to hold the feature-gated logic, but the existing implementation already doesn't do that. Given the ugly reality of this I will try to hold to the spirit of that goal and you can tell me what you think. I'll ping you when I push another commit this afternoon.

bswartz commented:
That last commit represents the changes to DropDisabledFields() and more tests. I still haven't reworded the comments/docs for the new core field. That will come next.

bswartz commented:
The most recent commit also updates the doc comments, and should cover all of the feedback so far. Please let me know if the rewrite of DropDisabledFields() is clear enough and follows the spirit of the feature-gate rule. The original logic was flawed so this new version is not only more readable but also more functional.

thockin commented on pkg/api/persistentvolumeclaim/util.go:
```diff
@@ -39,26 +65,21 @@ func dataSourceInUse(oldPVCSpec *core.PersistentVolumeClaimSpec) bool {
 	return false
 }
 
 func dataSourceIsEnabled(pvcSpec *core.PersistentVolumeClaimSpec) bool {
```
IsEnabled is now a total misnomer.

bswartz replied:
DropDisabled is called by Create and Update, and Normalize is called by Create and Get/List. I'm not opposed to more refactoring, but I think we'll end up with a 3rd function and the unit tests will need a complete rewrite as a result.

I also agree that the existing method names suck, but there's a tension between wanting to clean up bad code and wanting to avoid fixing stuff that's not broken.

thockin replied:
It *is* broken, it's just broken in human-space not compiler-space :)

thockin reviewed (commented):
This actually pushed me farther down the "move it" road.  IIUC, updates of PVC can't change the data source, so having this logic in DropDisabled, which has an "old" and "new" param is extra confusing.  Putting it in Normalize, which is only called on Create will make it that much more obvious what the heck is going on.

See, there's a story that you need to tell here.  3, 4, 5 years from now, someone will be staring at this and asking "WTF?!", running `git blame` and muttering about you.  Very possibly that someone will be you. 

I think you should move it, and tell the whole story in Normalize.  You are swimming upstream.

If you need to merge this and followup, I could buy that, but I REALLY think we will regret not doing it right.

bswartz commented:
Okay I have split most of the code in DropDisabledFields into another function, so Drop just handles the feature gate dropping. The new function is called EnforceDataSourceBackwardsCompatibility (long, but hopefully not confusing). I don't think it's possible to combine this new function with NormalizeDataSources, because Normalize is also used on the Get/List code path.

Lots of unit tests needed updating. And also I realized that updates needed some work so I fixed that code path too. In particular we have to be careful not to break any updates which would have previously worked, given that the validation code prevents mutations of PVC specs, and we're playing a lot of games with automatically modifying these fields.

I still want to add a raft of E2E tests that cover the combination of all the changes, to try to smoke out any corner cases, especially around combinations of updates and changing of the feature gate from off to on or on to off.

bswartz commented:
/retest

thockin commented on pkg/api/persistentvolumeclaim/util.go:
```diff
@@ -29,34 +29,25 @@ const (
+// to maintain backwards compatibility with old behavior.
+// See KEP 1495 for details.
+// Specifically, if this is an update of a PVC with no data source, or a creation of a new PVC,
+// and either the AnyVolumeDataSource feature gate is disabled or the dataSourceRef field is
```
Nitty: I wouldn't mention the feature gate here - it's not actually checked and it is irrelevant.  The field is either in use or not.  The gate is handled elsewhere.

thockin replied:
As discussed in slack - this is the right place to tell the full story about why it is called on both create and update paths, and how it handles back-rev clients

thockin reviewed (commented):
LGTM from API POV

/approve

bswartz commented:
I'm going to squash the commits into one so it can merge

bswartz commented:
Squashed

xing-yang commented:
/lgtm

xing-yang commented:
/milestone v1.22

caesarxuchao commented:
/triage accepted
